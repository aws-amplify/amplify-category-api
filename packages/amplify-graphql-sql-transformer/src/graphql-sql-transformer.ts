import {
  DirectiveWrapper,
  generateGetArgumentsInput,
  getResourceNamesForStrategy,
  InvalidDirectiveError,
  isSqlStrategy,
  MappingTemplate,
  TransformerPluginBase,
} from '@aws-amplify/graphql-transformer-core';
import { SqlDirective } from '@aws-amplify/graphql-directives';
import { TransformerContextProvider, TransformerSchemaVisitStepContextProvider } from '@aws-amplify/graphql-transformer-interfaces';
import * as cdk from 'aws-cdk-lib';
import {
  obj,
  str,
  ref,
  printBlock,
  compoundExpression,
  Expression,
  set,
  methodCall,
  ifElse,
  toJson,
  iff,
  notEquals,
  not,
} from 'graphql-mapping-template';
import { ResolverResourceIDs, ResourceConstants } from 'graphql-transformer-common';
import { DirectiveNode, ObjectTypeDefinitionNode, InterfaceTypeDefinitionNode, FieldDefinitionNode } from 'graphql';

type SqlDirectiveConfiguration = {
  statement: string | undefined;
  reference: string | undefined;
  resolverTypeName: string;
  resolverFieldName: string;
};

const SQL_DIRECTIVE_STACK = 'CustomSQLStack';

export class SqlTransformer extends TransformerPluginBase {
  private sqlDirectiveFields: Map<FieldDefinitionNode, SqlDirectiveConfiguration[]> = new Map();

  constructor() {
    super('amplify-sql-transformer', SqlDirective.definition);
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    ctx: TransformerSchemaVisitStepContextProvider,
  ): void => {
    if (parent.name.value !== 'Query' && parent.name.value !== 'Mutation') {
      throw new InvalidDirectiveError(
        `@sql directive can only be used on Query or Mutation types. Check type "${parent.name.value}" and field "${definition.name.value}".`,
      );
    }

    const directiveWrapped = new DirectiveWrapper(directive);

    // This should never happen -- it will be picked up during schema validation
    if (
      !directive?.arguments?.find((arg) => arg.name.value === 'statement') &&
      !directive?.arguments?.find((arg) => arg.name.value === 'reference')
    ) {
      throw new InvalidDirectiveError(
        `@sql directive must have either a 'statement' or a 'reference' argument. Check type "${parent.name.value}" and field "${definition.name.value}".`,
      );
    }

    const args = directiveWrapped.getArguments(
      {
        resolverTypeName: parent.name.value,
        resolverFieldName: definition.name.value,
      } as SqlDirectiveConfiguration,
      generateGetArgumentsInput(ctx.transformParameters),
    );
    let resolver = this.sqlDirectiveFields.get(definition);

    if (resolver === undefined) {
      resolver = [];
      this.sqlDirectiveFields.set(definition, resolver);
    }

    resolver.push(args);
  };

  generateResolvers = (context: TransformerContextProvider): void => {
    if (this.sqlDirectiveFields.size === 0) {
      return;
    }

    const stack: cdk.Stack = context.stackManager.createStack(SQL_DIRECTIVE_STACK);
    const env = context.synthParameters.amplifyEnvironmentName;

    stack.templateOptions.templateFormatVersion = '2010-09-09';
    stack.templateOptions.description = 'An auto-generated nested stack for the @sql directive.';

    new cdk.CfnCondition(stack, ResourceConstants.CONDITIONS.HasEnvironmentParameter, {
      expression: cdk.Fn.conditionNot(cdk.Fn.conditionEquals(env, ResourceConstants.NONE)),
    });

    this.sqlDirectiveFields.forEach((resolverFns) => {
      resolverFns.forEach((config) => {
        const typeName = config.resolverTypeName;
        const fieldName = config.resolverFieldName;
        const strategy = context.sqlDirectiveDataSourceStrategies?.find((css) => css.typeName === typeName && css.fieldName === fieldName);
        if (!strategy) {
          throw new Error(`Could not find custom SQL strategy for ${typeName}.${fieldName}`);
        }

        const resourceNames = getResourceNamesForStrategy(strategy.strategy);
        const dataSourceId = resourceNames.sqlLambdaDataSource;
        const dataSource = context.api.host.getDataSource(dataSourceId);

        const statement = getStatement(config, strategy.customSqlStatements);
        const resolverResourceId = ResolverResourceIDs.ResolverResourceID(config.resolverTypeName, config.resolverFieldName);
        const resolver = context.resolvers.generateQueryResolver(
          config.resolverTypeName,
          config.resolverFieldName,
          resolverResourceId,
          dataSource as any,
          MappingTemplate.s3MappingTemplateFromString(
            generateSqlLambdaRequestTemplate(statement, 'RAW_SQL', config.resolverFieldName),
            `${config.resolverTypeName}.${config.resolverFieldName}.req.vtl`,
          ),
          MappingTemplate.s3MappingTemplateFromString(
            generateSqlLambdaResponseMappingTemplate(),
            `${config.resolverTypeName}.${config.resolverFieldName}.res.vtl`,
          ),
        );

        resolver.addToSlot(
          'postAuth',
          MappingTemplate.s3MappingTemplateFromString(
            generateAuthExpressionForSandboxMode(context.transformParameters.sandboxModeEnabled),
            `${config.resolverTypeName}.${config.resolverFieldName}.{slotName}.{slotIndex}.req.vtl`,
          ),
        );
        resolver.setScope(context.stackManager.getScopeFor(resolverResourceId, SQL_DIRECTIVE_STACK));
        context.resolvers.addResolver(config.resolverTypeName, config.resolverFieldName, resolver);
      });
    });
  };
}

const generateAuthExpressionForSandboxMode = (enabled: boolean): string => {
  let exp;
  const API_KEY = 'API Key Authorization';

  if (enabled) exp = iff(notEquals(methodCall(ref('util.authType')), str(API_KEY)), methodCall(ref('util.unauthorized')));
  else exp = methodCall(ref('util.unauthorized'));

  return printBlock(`Sandbox Mode ${enabled ? 'Enabled' : 'Disabled'}`)(
    compoundExpression([iff(not(ref('ctx.stash.get("hasAuth")')), exp), toJson(obj({}))]),
  );
};

const getStatementFromStatementAttribute = (config: SqlDirectiveConfiguration): string => {
  const statement = config.statement;
  if (statement === undefined || statement.trim().length === 0) {
    throw new InvalidDirectiveError(
      `@sql directive 'statement' argument must not be empty. Check type "${config.resolverTypeName}" and field "${config.resolverFieldName}".`,
    );
  }
  return statement;
};

const getStatementFromReferenceAttribute = (config: SqlDirectiveConfiguration, customQueries?: Record<string, string>): string => {
  if (!config.reference || !customQueries || !customQueries[config.reference]) {
    throw new InvalidDirectiveError(
      `The ${config.resolverTypeName} field "${config.resolverFieldName}" references a custom SQL statement "${config.reference}" that ` +
        `doesn't exist. Verify that "${config.reference}" is a key in the customSqlStatements property.`,
    );
  }
  return customQueries[config.reference];
};

const getStatement = (config: SqlDirectiveConfiguration, customQueries?: Record<string, string>): string => {
  if (config.reference && config.statement) {
    throw new InvalidDirectiveError(
      `@sql directive can have either a 'statement' or a 'reference' argument but not both. Check type "${config.resolverTypeName}" and field "${config.resolverFieldName}".`,
    );
  }

  if (typeof config.statement === 'string') {
    return getStatementFromStatementAttribute(config);
  }

  if (typeof config.reference === 'string') {
    return getStatementFromReferenceAttribute(config, customQueries);
  }

  // This should never happen -- it will be picked up during schema validation -- but we'll be defensive and ensure the type safety of the
  // function
  throw new InvalidDirectiveError(
    `@sql directive must have either a 'statement' or a 'reference' argument. Check type "${config.resolverTypeName}" and field "${config.resolverFieldName}".`,
  );
};

export const generateSqlLambdaRequestTemplate = (statement: string, operation: string, operationName: string): string => {
  return printBlock('Invoke RDS Lambda data source')(
    compoundExpression([
      set(ref('lambdaInput'), obj({})),
      set(ref('lambdaInput.parameters'), obj({})),
      set(ref('lambdaInput.statement'), str(statement)),
      set(ref('lambdaInput.operation'), str(operation)),
      set(ref('lambdaInput.operationName'), str(operationName)),
      set(ref('lambdaInput.parameters'), methodCall(ref('util.defaultIfNull'), ref('context.arguments'), obj({}))),
      obj({
        version: str('2018-05-29'),
        operation: str('Invoke'),
        payload: methodCall(ref('util.toJson'), ref('lambdaInput')),
      }),
    ]),
  );
};

export const generateSqlLambdaResponseMappingTemplate = (): string => {
  const statements: Expression[] = [];

  statements.push(
    ifElse(ref('ctx.error'), methodCall(ref('util.error'), ref('ctx.error.message'), ref('ctx.error.type')), toJson(ref('ctx.result'))),
  );

  return printBlock('ResponseTemplate')(compoundExpression(statements));
};
