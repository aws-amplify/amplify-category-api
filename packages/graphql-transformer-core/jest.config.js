const baseConfig = require('../../jest.config.base.js'); // eslint-disable-line @typescript-eslint/no-var-requires

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      branches: 65,
      functions: 33,
      lines: 47,
    },
  },
};
