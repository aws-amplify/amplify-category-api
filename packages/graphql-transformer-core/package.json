{
  "name": "graphql-transformer-core",
  "version": "8.2.4",
  "description": "A framework to transform from GraphQL SDL to AWS cloudFormation.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-category-api.git",
    "directory": "packages/graphql-transformer-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "cloudformation",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "build": "rimraf ./lib ./tsconfig.tsbuildinfo && tsc",
    "watch": "tsc -w",
    "clean": "rimraf ./lib",
    "extract-api": "ts-node ../../scripts/extract-api.ts"
  },
  "dependencies": {
    "cloudform-types": "^4.2.0",
    "deep-diff": "^1.0.2",
    "fs-extra": "^8.1.0",
    "glob": "^10.3.0",
    "graphql": "^15.5.0",
    "graphql-transformer-common": "4.27.1",
    "lodash": "^4.17.21"
  },
  "devDependencies": {
    "@aws-amplify/amplify-cli-core": "^4.2.10",
    "@types/fs-extra": "^8.0.1",
    "@types/node": "^18.17.0",
    "mock-fs": "5.2.0",
    "@types/mock-fs": "4.13.4"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true,
    "coverageProvider": "v8",
    "coverageThreshold": {
      "global": {
        "branches": 67,
        "functions": 33,
        "lines": 48
      }
    },
    "coverageReporters": [
      "clover",
      "text"
    ]
  }
}
