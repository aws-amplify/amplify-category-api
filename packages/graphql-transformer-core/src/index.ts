import './polyfills/Object.assign';
import { print } from 'graphql';
import { DeploymentResources } from './DeploymentResources';
import { TransformerContext, MappingParameters } from './TransformerContext';
import { Transformer } from './Transformer';
import { ITransformer } from './ITransformer';
import { GraphQLTransform } from './GraphQLTransform';
import { collectDirectiveNames, collectDirectivesByType, collectDirectivesByTypeNames } from './collectDirectives';
import { stripDirectives } from './stripDirectives';
import {
  buildProject as buildAPIProject,
  uploadDeployment as uploadAPIProject,
  migrateAPIProject,
  revertAPIMigration,
  CLOUDFORMATION_FILE_NAME,
  PARAMETERS_FILE_NAME,
} from './util/amplifyUtils';
import {
  readSchema as readProjectSchema,
  loadProject as readProjectConfiguration,
  loadConfig as readTransformerConfiguration,
  writeConfig as writeTransformerConfiguration,
} from './util/transformConfig';
import { EXTRA_DIRECTIVES_DOCUMENT } from './validation';

export * from './errors';
export * from './util';
export { getTableNameForModel } from './tableNameMap';

/**
 * Returns the set of directives that are supported by AppSync service
 */
export function getAppSyncServiceExtraDirectives(): string {
  return print(EXTRA_DIRECTIVES_DOCUMENT);
}
export { FeatureFlagProvider } from './FeatureFlags';

export {
  GraphQLTransform,
  TransformerContext,
  MappingParameters,
  Transformer,
  ITransformer,
  collectDirectiveNames,
  collectDirectivesByType,
  collectDirectivesByTypeNames,
  stripDirectives,
  buildAPIProject,
  migrateAPIProject,
  uploadAPIProject,
  readProjectSchema,
  readProjectConfiguration,
  readTransformerConfiguration,
  writeTransformerConfiguration,
  revertAPIMigration,
  DeploymentResources,
  CLOUDFORMATION_FILE_NAME,
  PARAMETERS_FILE_NAME,
};
// No-op change to trigger publish
