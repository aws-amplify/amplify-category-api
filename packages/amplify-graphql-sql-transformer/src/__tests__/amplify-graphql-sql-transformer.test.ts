import { constructDataSourceStrategies, validateModelSchema } from '@aws-amplify/graphql-transformer-core';
import { parse } from 'graphql';
import { TestTransformParameters, mockSqlDataSourceStrategy, testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { SqlTransformer } from '../graphql-sql-transformer';

describe('sql directive tests', () => {
  const mySqlStrategy = mockSqlDataSourceStrategy();

  it('should compile happy case with statement argument', () => {
    const doc = /* GraphQL */ `
      type Query {
        calculateTaxRate(zip: String): Int @sql(statement: "SELECT * FROM TAXRATE WHERE ZIP = :zip")
      }
    `;

    const out = testTransform({
      schema: doc,
      transformers: [new ModelTransformer(), new SqlTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
      sqlDirectiveDataSourceStrategies: [
        {
          typeName: 'Query',
          fieldName: 'calculateTaxRate',
          strategy: mySqlStrategy,
        },
      ],
    });
    expect(out).toBeDefined();
    expect(out.schema).toBeDefined();
    const definition = out.schema;
    expect(definition).toBeDefined();

    const parsed = parse(definition);
    validateModelSchema(parsed);

    expect(out.resolvers).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.req.vtl']).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.res.vtl']).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Query.calculateTaxRate.res.vtl']).toMatchSnapshot();
  });

  it('should compile happy case with reference argument', () => {
    const doc = /* GraphQL */ `
      type Query {
        calculateTaxRate(zip: String): Int @sql(reference: "calculate-tax")
      }
    `;

    const out = testTransform({
      schema: doc,
      transformers: [new ModelTransformer(), new SqlTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
      sqlDirectiveDataSourceStrategies: [
        {
          typeName: 'Query',
          fieldName: 'calculateTaxRate',
          strategy: mySqlStrategy,
          customSqlStatements: {
            'calculate-tax': 'SELECT * FROM TAXRATE WHERE ZIP = :zip',
          },
        },
      ],
    });
    expect(out).toBeDefined();
    expect(out.schema).toBeDefined();
    const definition = out.schema;
    expect(definition).toBeDefined();

    const parsed = parse(definition);
    validateModelSchema(parsed);

    expect(out.resolvers).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.req.vtl']).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.res.vtl']).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Query.calculateTaxRate.res.vtl']).toMatchSnapshot();
  });

  it('should throw error if incorrect reference argument', () => {
    const doc = /* GraphQL */ `
      type Query {
        calculateTaxRate(zip: String): Int @sql(reference: "calculate-tax")
      }
    `;

    const transformConfig: TestTransformParameters = {
      schema: doc,
      transformers: [new ModelTransformer(), new SqlTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
      sqlDirectiveDataSourceStrategies: [
        {
          typeName: 'Query',
          fieldName: 'calculateTaxRate',
          strategy: mySqlStrategy,
          customSqlStatements: {
            'incorrect-reference-name': 'SELECT * FROM TAXRATE WHERE ZIP = :zip',
          },
        },
      ],
    };

    expect(() => testTransform(transformConfig)).toThrowError(
      'The Query field "calculateTaxRate" references a custom SQL statement "calculate-tax" that doesn\'t exist. Verify that "calculate-tax" is a key in the customSqlStatements property.',
    );
  });

  it('should throw error if both statement and argument provided', () => {
    const doc = /* GraphQL */ `
      type Query {
        calculateTaxRate(zip: String): Int @sql(statement: "SELECT * FROM TAXRATE WHERE ZIP = :zip", reference: "calculate-tax")
      }
    `;

    const customQueries = new Map<string, string>();
    customQueries.set('calculate-tax', 'SELECT * FROM TAXRATE WHERE ZIP = :zip');

    const transformConfig: TestTransformParameters = {
      schema: doc,
      transformers: [new ModelTransformer(), new SqlTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
      sqlDirectiveDataSourceStrategies: [
        {
          typeName: 'Query',
          fieldName: 'calculateTaxRate',
          strategy: mySqlStrategy,
          customSqlStatements: {
            'calculate-tax': 'SELECT * FROM TAXRATE WHERE ZIP = :zip',
          },
        },
      ],
    };

    expect(() => testTransform(transformConfig)).toThrowError(
      '@sql directive can have either a \'statement\' or a \'reference\' argument but not both. Check type "Query" and field "calculateTaxRate".',
    );
  });

  it('should throw error if neither statement and argument provided', () => {
    const doc = /* GraphQL */ `
      type Query {
        calculateTaxRate(zip: String): Int @sql
      }
    `;

    const customQueries = new Map<string, string>();
    customQueries.set('calculate-tax', 'SELECT * FROM TAXRATE WHERE ZIP = :zip');

    const transformConfig: TestTransformParameters = {
      schema: doc,
      transformers: [new ModelTransformer(), new SqlTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
      sqlDirectiveDataSourceStrategies: [
        {
          typeName: 'Query',
          fieldName: 'calculateTaxRate',
          strategy: mySqlStrategy,
          customSqlStatements: {
            'calculate-tax': 'SELECT * FROM TAXRATE WHERE ZIP = :zip',
          },
        },
      ],
    };

    expect(() => testTransform(transformConfig)).toThrowError(
      '@sql directive must have either a \'statement\' or a \'reference\' argument. Check type "Query" and field "calculateTaxRate".',
    );
  });

  it('should throw error if statement is empty', () => {
    const doc = /* GraphQL */ `
      type Query {
        calculateTaxRate(zip: String): Int @sql(statement: "")
      }
    `;

    const transformConfig: TestTransformParameters = {
      schema: doc,
      transformers: [new ModelTransformer(), new SqlTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
      sqlDirectiveDataSourceStrategies: [
        {
          typeName: 'Query',
          fieldName: 'calculateTaxRate',
          strategy: mySqlStrategy,
        },
      ],
    };

    expect(() => testTransform(transformConfig)).toThrowError(
      '@sql directive \'statement\' argument must not be empty. Check type "Query" and field "calculateTaxRate".',
    );
  });

  it('throws an error if invoked with the wrong type', () => {
    const doc = /* GraphQL */ `
      type Todo {
        calculateTaxRate(zip: String): Int @sql(statement: "SELECT * FROM TAXRATE WHERE ZIP = :zip")
      }
    `;

    const transformConfig: TestTransformParameters = {
      schema: doc,
      transformers: [new ModelTransformer(), new SqlTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
      sqlDirectiveDataSourceStrategies: [
        {
          typeName: 'Todo' as any,
          fieldName: 'calculateTaxRate',
          strategy: mySqlStrategy,
        },
      ],
    };
    expect(() => testTransform(transformConfig)).toThrowError(
      '@sql directive can only be used on Query or Mutation types. Check type "Todo" and field "calculateTaxRate".',
    );
  });

  it('successfully processes a schema with only custom SQL', () => {
    const doc = /* GraphQL */ `
      type Query {
        calculateTaxRate(zip: String): Int @sql(statement: "SELECT * FROM TAXRATE WHERE ZIP = :zip")
      }
    `;

    const transformConfig: TestTransformParameters = {
      schema: doc,
      transformers: [new ModelTransformer(), new SqlTransformer()],
      dataSourceStrategies: {},
      sqlDirectiveDataSourceStrategies: [
        {
          typeName: 'Query',
          fieldName: 'calculateTaxRate',
          strategy: mySqlStrategy,
        },
      ],
    };

    const out = testTransform(transformConfig);
    expect(out).toBeDefined();
    expect(out.schema).toBeDefined();
    const definition = out.schema;
    expect(definition).toBeDefined();

    const parsed = parse(definition);
    validateModelSchema(parsed);

    expect(out.resolvers).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.req.vtl']).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.res.vtl']).toBeDefined();
    expect(out.resolvers['Query.calculateTaxRate.req.vtl']).toMatchSnapshot();
    expect(out.resolvers['Query.calculateTaxRate.res.vtl']).toMatchSnapshot();
  });

  describe('Post auth resolver', () => {
    it('Renders post auth resolver if sandbox is enabled and iam access disabled', () => {
      const doc = /* GraphQL */ `
        type Query {
          calculateTaxRate(zip: String): Int @sql(statement: "SELECT * FROM TAXRATE WHERE ZIP = :zip")
        }
      `;

      const out = testTransform({
        schema: doc,
        transformers: [new ModelTransformer(), new SqlTransformer()],
        dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
        transformParameters: {
          sandboxModeEnabled: true,
        },
        synthParameters: {
          enableIamAccess: false,
        },
        sqlDirectiveDataSourceStrategies: [
          {
            typeName: 'Query',
            fieldName: 'calculateTaxRate',
            strategy: mySqlStrategy,
          },
        ],
      });
      expect(out).toBeDefined();
      expect(out.resolvers).toBeDefined();
      expect(out.resolvers['Query.calculateTaxRate.postAuth.1.req.vtl']).toBeDefined();
      expect(out.resolvers['Query.calculateTaxRate.postAuth.1.req.vtl']).toMatchSnapshot();
    });

    it('Renders post auth resolver if sandbox is enabled and iam access enabled', () => {
      const doc = /* GraphQL */ `
        type Query {
          calculateTaxRate(zip: String): Int @sql(statement: "SELECT * FROM TAXRATE WHERE ZIP = :zip")
        }
      `;

      const out = testTransform({
        schema: doc,
        transformers: [new ModelTransformer(), new SqlTransformer()],
        dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
        transformParameters: {
          sandboxModeEnabled: true,
        },
        synthParameters: {
          enableIamAccess: true,
        },
        sqlDirectiveDataSourceStrategies: [
          {
            typeName: 'Query',
            fieldName: 'calculateTaxRate',
            strategy: mySqlStrategy,
          },
        ],
      });
      expect(out).toBeDefined();
      expect(out.resolvers).toBeDefined();
      expect(out.resolvers['Query.calculateTaxRate.postAuth.1.req.vtl']).toBeDefined();
      expect(out.resolvers['Query.calculateTaxRate.postAuth.1.req.vtl']).toMatchSnapshot();
    });

    it('Renders post auth resolver if sandbox is disabled and iam access enabled', () => {
      const doc = /* GraphQL */ `
        type Query {
          calculateTaxRate(zip: String): Int @sql(statement: "SELECT * FROM TAXRATE WHERE ZIP = :zip")
        }
      `;

      const out = testTransform({
        schema: doc,
        transformers: [new ModelTransformer(), new SqlTransformer()],
        dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
        transformParameters: {
          sandboxModeEnabled: false,
        },
        synthParameters: {
          enableIamAccess: true,
        },
        sqlDirectiveDataSourceStrategies: [
          {
            typeName: 'Query',
            fieldName: 'calculateTaxRate',
            strategy: mySqlStrategy,
          },
        ],
      });
      expect(out).toBeDefined();
      expect(out.resolvers).toBeDefined();
      expect(out.resolvers['Query.calculateTaxRate.postAuth.1.req.vtl']).toBeDefined();
      expect(out.resolvers['Query.calculateTaxRate.postAuth.1.req.vtl']).toMatchSnapshot();
    });

    it('Renders post auth resolver if sandbox is disabled and iam access disabled', () => {
      const doc = /* GraphQL */ `
        type Query {
          calculateTaxRate(zip: String): Int @sql(statement: "SELECT * FROM TAXRATE WHERE ZIP = :zip")
        }
      `;

      const out = testTransform({
        schema: doc,
        transformers: [new ModelTransformer(), new SqlTransformer()],
        dataSourceStrategies: constructDataSourceStrategies(doc, mySqlStrategy),
        transformParameters: {
          sandboxModeEnabled: false,
        },
        synthParameters: {
          enableIamAccess: false,
        },
        sqlDirectiveDataSourceStrategies: [
          {
            typeName: 'Query',
            fieldName: 'calculateTaxRate',
            strategy: mySqlStrategy,
          },
        ],
      });
      expect(out).toBeDefined();
      expect(out.resolvers).toBeDefined();
      expect(out.resolvers['Query.calculateTaxRate.postAuth.1.req.vtl']).toBeDefined();
      expect(out.resolvers['Query.calculateTaxRate.postAuth.1.req.vtl']).toMatchSnapshot();
    });
  });
});
