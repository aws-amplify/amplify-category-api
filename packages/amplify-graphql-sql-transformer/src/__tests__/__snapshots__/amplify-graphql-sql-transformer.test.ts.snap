// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql directive tests Post auth resolver Renders post auth resolver if sandbox is disabled and iam access disabled 1`] = `
"## [Start] Sandbox Mode Disabled. **
#if( !$ctx.stash.get(\\"hasAuth\\") )
  $util.unauthorized()
#end
$util.toJson({})
## [End] Sandbox Mode Disabled. **"
`;

exports[`sql directive tests Post auth resolver Renders post auth resolver if sandbox is disabled and iam access enabled 1`] = `
"## [Start] Sandbox Mode Disabled. **
#if( !$ctx.stash.get(\\"hasAuth\\") )
  #if( $util.authType() == \\"IAM Authorization\\" && $util.isNull($ctx.identity.cognitoIdentityPoolId) && $util.isNull($ctx.identity.cognitoIdentityId) )
    #return($util.toJson({}))
  #end
  $util.unauthorized()
#end
$util.toJson({})
## [End] Sandbox Mode Disabled. **"
`;

exports[`sql directive tests Post auth resolver Renders post auth resolver if sandbox is enabled and iam access disabled 1`] = `
"## [Start] Sandbox Mode Enabled. **
#if( !$ctx.stash.get(\\"hasAuth\\") )
  #if( $util.authType() == \\"API Key Authorization\\" )
    #return($util.toJson({}))
  #end
  $util.unauthorized()
#end
$util.toJson({})
## [End] Sandbox Mode Enabled. **"
`;

exports[`sql directive tests Post auth resolver Renders post auth resolver if sandbox is enabled and iam access enabled 1`] = `
"## [Start] Sandbox Mode Enabled. **
#if( !$ctx.stash.get(\\"hasAuth\\") )
  #if( $util.authType() == \\"API Key Authorization\\" )
    #return($util.toJson({}))
  #end
  #if( $util.authType() == \\"IAM Authorization\\" && $util.isNull($ctx.identity.cognitoIdentityPoolId) && $util.isNull($ctx.identity.cognitoIdentityId) )
    #return($util.toJson({}))
  #end
  $util.unauthorized()
#end
$util.toJson({})
## [End] Sandbox Mode Enabled. **"
`;

exports[`sql directive tests should compile happy case with reference argument 1`] = `
"## [Start] Invoke RDS Lambda data source. **
#set( $lambdaInput = {} )
#set( $lambdaInput.parameters = {} )
#set( $lambdaInput.statement = \\"SELECT * FROM TAXRATE WHERE ZIP = :zip\\" )
#set( $lambdaInput.operation = \\"RAW_SQL\\" )
#set( $lambdaInput.operationName = \\"calculateTaxRate\\" )
#set( $lambdaInput.parameters = $util.defaultIfNull($context.arguments, {}) )
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson($lambdaInput)
}
## [End] Invoke RDS Lambda data source. **"
`;

exports[`sql directive tests should compile happy case with reference argument 2`] = `
"## [Start] ResponseTemplate. **
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#else
  $util.toJson($ctx.result)
#end
## [End] ResponseTemplate. **"
`;

exports[`sql directive tests should compile happy case with statement argument 1`] = `
"## [Start] Invoke RDS Lambda data source. **
#set( $lambdaInput = {} )
#set( $lambdaInput.parameters = {} )
#set( $lambdaInput.statement = \\"SELECT * FROM TAXRATE WHERE ZIP = :zip\\" )
#set( $lambdaInput.operation = \\"RAW_SQL\\" )
#set( $lambdaInput.operationName = \\"calculateTaxRate\\" )
#set( $lambdaInput.parameters = $util.defaultIfNull($context.arguments, {}) )
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson($lambdaInput)
}
## [End] Invoke RDS Lambda data source. **"
`;

exports[`sql directive tests should compile happy case with statement argument 2`] = `
"## [Start] ResponseTemplate. **
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#else
  $util.toJson($ctx.result)
#end
## [End] ResponseTemplate. **"
`;

exports[`sql directive tests successfully processes a schema with only custom SQL 1`] = `
"## [Start] Invoke RDS Lambda data source. **
#set( $lambdaInput = {} )
#set( $lambdaInput.parameters = {} )
#set( $lambdaInput.statement = \\"SELECT * FROM TAXRATE WHERE ZIP = :zip\\" )
#set( $lambdaInput.operation = \\"RAW_SQL\\" )
#set( $lambdaInput.operationName = \\"calculateTaxRate\\" )
#set( $lambdaInput.parameters = $util.defaultIfNull($context.arguments, {}) )
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson($lambdaInput)
}
## [End] Invoke RDS Lambda data source. **"
`;

exports[`sql directive tests successfully processes a schema with only custom SQL 2`] = `
"## [Start] ResponseTemplate. **
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#else
  $util.toJson($ctx.result)
#end
## [End] ResponseTemplate. **"
`;
